fem/eltrans/eltrans_basis.hpp

```diff
@@ -0,0 +1,251 @@
+// Copyright (c) 2010-2025, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_ELTRANS_BASIS
+#define MFEM_ELTRANS_BASIS
+
+#include "../../general/forall.hpp"
+
+#include "../geom.hpp"
+
+// this file contains utilities for computing nodal basis functions and their
+// derivatives in device kernels
+
+namespace mfem
+{
+
+namespace eltrans
+{
+
+/// Various utilities for working with different element geometries
+template <int GeomType> struct GeometryUtils;
+
+template <> struct GeometryUtils<Geometry::SEGMENT>
+{
+   static constexpr MFEM_HOST_DEVICE int Dimension() { return 1; }
+
+   /// @b true if the given point x in ref space is inside the element
+   static bool MFEM_HOST_DEVICE inside(real_t x) { return x >= 0 && x <= 1; }
+
+   /// @b Bound the reference coordinate @a x += dx to be inside the segment.
+   /// @a dx is updated to be dx = project(x+dx) - x
+   static bool MFEM_HOST_DEVICE project(real_t &x, real_t &dx)
+   {
+      real_t tmp = x;
+      x += dx;
+      if (x < 0)
+      {
+         x = 0;
+         dx = x - tmp;
+         return true;
+      }
+      if (x > 1)
+      {
+         x = 1;
+         dx = x - tmp;
+         return true;
+      }
+      return false;
+   }
+};
+
+template <> struct GeometryUtils<Geometry::SQUARE>
+{
+   static constexpr MFEM_HOST_DEVICE int Dimension() { return 2; }
+   /// @b true if the given point (x,y) in ref space is inside the element
+   static bool MFEM_HOST_DEVICE inside(real_t x, real_t y)
+   {
+      return (x >= 0) && (x <= 1) && (y >= 0) && (y <= 1);
+   }
+
+   /// @b Bound the reference coordinate @a (x,y) += (dx,dy) to be inside the
+   /// square.
+   /// @a dx and @a dy are updated to be (dx,dy) = project(x+dx,y+dy) - (x,y)
+   static bool MFEM_HOST_DEVICE project(real_t &x, real_t &y, real_t &dx,
+                                        real_t &dy)
+   {
+      bool res = false;
+      real_t tmp = x;
+      x += dx;
+      if (x < 0)
+      {
+         res = true;
+         x = 0;
+         dx = x - tmp;
+      }
+      else if (x > 1)
+      {
+         res = true;
+         x = 1;
+         dx = x - tmp;
+      }
+      tmp = y;
+      y += dy;
+      if (y < 0)
+      {
+         res = true;
+         y = 0;
+         dy = y - tmp;
+      }
+      else if (y > 1)
+      {
+         res = true;
+         y = 1;
+         dy = y - tmp;
+      }
+      return res;
```

Comment:
Minor:

Would it make sense to implement this in terms of the 1D function?
```suggestion
      const bool x_res = GeometryUtils<Geometry::SEGMENT>::project(x,dx);
      const bool y_res = GeometryUtils<Geometry::SEGMENT>::project(y,dy);
      return x_res || y_res;
```

and similarly for the `Geometry::CUBE` case.

Reply:
Yes, good suggestion.